File: js/142_LinkedListCycleII.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var detectCycle = function(head) {
    if(!head) return null;
    var fast = head;
    var slow = head;
    while(fast.next !== null && fast.next.next !== null){
        fast = fast.next.next;
        slow = slow.next;
        if(fast === slow){
            var detector = head;
            while(detector !== slow){
                slow = slow.next;
                detector = detector.next;
            }
            return detector;
        }
    }
    return null;
};

File: cpp/142_LinkedListCycleII.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if(!head) return NULL;
        auto fast = head;
        auto slow = head;
        while(fast->next != NULL && fast->next->next != NULL){
            fast = fast->next->next;
            slow = slow->next;
            if(fast == slow){
                auto detector = head;
                while(detector != slow){
                    detector = detector->next;
                    slow = slow->next;
                }
                return slow;
            }
        }
        return NULL;
    }
};